File: scripts/tinker_onboard.js
#!/usr/bin/env node
/**
 * Tinker onboarding via Playwright Node.js
 * Complete onboarding and extract API key
 */

const PLAYWRIGHT_DIR = "/home/openclaw/.npm/_npx/e41f203b7505f1fb/node_modules";
const { chromium } = require(`${PLAYWRIGHT_DIR}/playwright`);
const fs = require('fs');

const OUTPUT = '/tmp/tinker_result.json';

async function run() {
  const result = { success: false, api_key: null, error: null, step: 'start' };

  console.log('Launching headless Chromium...');
  const browser = await chromium.launch({
    headless: true,
    args: ['--no-sandbox', '--disable-setuid-sandbox', '--disable-dev-shm-usage']
  });

  const context = await browser.newContext({
    userAgent: 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36'
  });

  const page = await context.newPage();
  result.step = 'browser_ready';

  try {
    console.log('Navigating to Tinker /keys...');
    await page.goto('https://tinker-console.thinkingmachines.ai/keys', { timeout: 15000 });
    await page.waitForLoadState('networkidle', { timeout: 10000 });
    
    let url = page.url();
    console.log(`Current URL: ${url}`);
    result.step = `navigated: ${url}`;

    // Need to auth?
    if (url.includes('auth.thinkingmachines.ai') || url.includes('sign')) {
      console.log('Need to auth, trying Google OAuth...');
      result.step = 'need_auth';

      // Trigger Google OAuth
      const oauthUrl = 'https://auth.thinkingmachines.ai/api/login?provider=GoogleOAuth' +
        '&redirect_uri=https%3A%2F%2Ftinker-console.thinkingmachines.ai%2Fcallback' +
        '&client_id=client_01JT41MFTDNYP0RYJ9MF318GDF&source=signin';
      
      await page.goto(oauthUrl, { timeout: 15000 });
      await page.waitForLoadState('networkidle', { timeout: 10000 });
      url = page.url();
      console.log(`OAuth redirect: ${url}`);

      if (url.includes('accounts.google.com')) {
        // Try to use existing account
        const accountLink = page.locator('text=muchammadizzuddin@gmail.com');
        const accountCount = await accountLink.count();
        
        if (accountCount > 0) {
          console.log('Found account in chooser, clicking...');
          await accountLink.click();
          await page.waitForLoadState('networkidle', { timeout: 15000 });
        } else {
          // Type email
          console.log('Typing email...');
          await page.fill('input[type="email"]', 'muchammadizzuddin@gmail.com');
          await page.click('button:has-text("Next"), #identifierNext');
          await page.waitForLoadState('networkidle', { timeout: 10000 });
          
          url = page.url();
          console.log(`After email: ${url}`);
          
          // Check if password needed
          const pwInput = page.locator('input[type="password"]');
          if (await pwInput.count() > 0) {
            result.error = 'NEED_GOOGLE_PASSWORD';
            result.step = 'blocked_need_password';
            console.log('BLOCKED: Google password required');
            await browser.close();
            fs.writeFileSync(OUTPUT, JSON.stringify(result, null, 2));
            return;
          }
        }
        
        // Google consent screen?
        url = page.url();
        console.log(`After Google: ${url}`);
        const continueBtn = page.locator('button:has-text("Continue")');
        if (await continueBtn.count() > 0) {
          await continueBtn.click();
          await page.waitForLoadState('networkidle', { timeout: 15000 });
        }
      }
    }

    url = page.url();
    console.log(`After auth: ${url}`);

    // Onboarding?
    if (url.includes('onboarding')) {
      console.log('On onboarding, completing...');
      result.step = 'onboarding';
      
      // Check ToS
      const checkbox = page.locator('input[type="checkbox"]');
      if (await checkbox.count() > 0) {
        const isChecked = await checkbox.isChecked();
        if (!isChecked) {
          await checkbox.check();
          console.log('ToS accepted');
        }
      }
      
      // Click Continue
      const btn = page.locator('button:has-text("Continue")');
      if (await btn.count() > 0 && await btn.isEnabled()) {
        await btn.click();
        await page.waitForLoadState('networkidle', { timeout: 15000 });
        console.log(`After onboarding: ${page.url()}`);
      }
    }

    url = page.url();
    console.log(`Final URL: ${url}`);

    // Navigate to keys if not there yet
    if (!url.includes('keys')) {
      await page.goto('https://tinker-console.thinkingmachines.ai/keys', { timeout: 15000 });
      await page.waitForLoadState('networkidle', { timeout: 10000 });
      url = page.url();
      console.log(`Keys URL: ${url}`);
    }

    // Extract API key from page
    const content = await page.content();
    result.final_url = url;
    result.page_title = await page.title();
    
    // Look for API key patterns (Tinker keys usually start with "tm_")
    const keyPatterns = [
      /tm_[a-zA-Z0-9_-]{20,}/g,
      /sk-[a-zA-Z0-9_-]{20,}/g,
      /"key":\s*"([^"]{20,})"/g,
      /"api_key":\s*"([^"]{20,})"/g,
    ];
    
    for (const pattern of keyPatterns) {
      const matches = content.match(pattern);
      if (matches && matches.length > 0) {
        result.api_key = matches[0];
        console.log(`Found API key: ${result.api_key.substring(0, 20)}...`);
        break;
      }
    }

    // Try clicking "Create" or "Generate" key button if no key found
    if (!result.api_key) {
      const createBtn = page.locator('button:has-text("Create"), button:has-text("Generate"), button:has-text("New"), button:has-text("+ ")');
      if (await createBtn.count() > 0) {
        console.log('Clicking create key button...');
        await createBtn.first().click();
        await page.waitForLoadState('networkidle', { timeout: 10000 });
        
        // Get new content
        const newContent = await page.content();
        for (const pattern of keyPatterns) {
          const matches = newContent.match(pattern);
          if (matches && matches.length > 0) {
            result.api_key = matches[0];
            console.log(`Found API key after create: ${result.api_key.substring(0, 20)}...`);
            break;
          }
        }
        
        // Also check text visible on page
        const keyText = await page.locator('[class*="key"], [class*="token"], [class*="api"]').allTextContents();
        console.log('Key elements text:', keyText.slice(0, 5));
        result.key_elements = keyText.slice(0, 5);
      }
    }

    result.success = !url.includes('auth') && !url.includes('sign');
    result.page_preview = content.substring(0, 3000);

  } catch (e) {
    result.error = e.message;
    console.error('Error:', e.message);
    try {
      result.final_url = page.url();
      result.page_preview = (await page.content()).substring(0, 2000);
    } catch {}
  }

  await browser.close();
  
  fs.writeFileSync(OUTPUT, JSON.stringify(result, null, 2));
  console.log('\n=== RESULT ===');
  console.log(JSON.stringify(result, null, 2));
}

run().catch(e => {
  console.error('Fatal:', e);
  fs.writeFileSync(OUTPUT, JSON.stringify({ success: false, error: e.message }, null, 2));
  process.exit(1);
});

File: scripts/tinker_onboard.py
#!/usr/bin/env python3
"""
Tinker onboarding via Playwright headless
Uses cookies from existing browser profile
"""
import json, os, shutil, sqlite3, tempfile, time
from pathlib import Path

PROFILE_DIR = "/home/openclaw/.openclaw/browser/openclaw/user-data"
COOKIES_DB = f"{PROFILE_DIR}/Default/Cookies"
OUTPUT_FILE = "/tmp/tinker_result.json"

def get_google_cookies():
    """Extract Google cookies from Chromium profile using secretstorage for decryption"""
    try:
        # Copy cookies db to temp (avoid lock issues)
        tmp = tempfile.mktemp(suffix=".db")
        shutil.copy2(COOKIES_DB, tmp)
        
        conn = sqlite3.connect(tmp)
        c = conn.cursor()
        # Get Google + Tinker cookies
        c.execute("""
            SELECT host_key, name, path, is_secure, expires_utc, encrypted_value
            FROM cookies 
            WHERE host_key LIKE '%.google.com' 
               OR host_key LIKE '%.thinkingmachines.ai'
               OR host_key LIKE '%.oauth.thinkingmachines.ai'
            ORDER BY host_key
        """)
        rows = c.fetchall()
        conn.close()
        os.unlink(tmp)
        print(f"Found {len(rows)} relevant cookies")
        return rows
    except Exception as e:
        print(f"Cookie extraction failed: {e}")
        return []

def decrypt_cookie(encrypted_value):
    """Decrypt Chrome cookie using Linux AES-256-CBC with secretstorage key"""
    try:
        import secretstorage, os
        os.environ['DBUS_SESSION_BUS_ADDRESS'] = 'unix:path=/run/user/1000/bus'
        bus = secretstorage.dbus_init()
        
        # Get Chrome Safe Storage key from login collection
        login_col = secretstorage.Collection(bus, '/org/freedesktop/secrets/collection/login')
        for item in login_col.get_all_items():
            attrs = item.get_attributes()
            if attrs.get('application') in ('chrome', 'Antigravity', 'chromium'):
                key = item.get_secret()
                break
        else:
            return None
        
        from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
        from cryptography.hazmat.primitives import hashes
        from cryptography.hazmat.backends import default_backend
        from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
        
        # Chrome on Linux: PBKDF2 with "peanuts" salt
        kdf = PBKDF2HMAC(
            algorithm=hashes.SHA1(),
            length=16,
            salt=b'saltysalt',
            iterations=1,
            backend=default_backend()
        )
        derived_key = kdf.derive(key)
        
        if encrypted_value[:3] == b'v10':
            encrypted_value = encrypted_value[3:]
            iv = b' ' * 16
            cipher = Cipher(algorithms.AES(derived_key), modes.CBC(iv), backend=default_backend())
            decryptor = cipher.decryptor()
            decrypted = decryptor.update(encrypted_value) + decryptor.finalize()
            # Remove PKCS7 padding
            pad_len = decrypted[-1]
            return decrypted[:-pad_len].decode('utf-8')
    except Exception as e:
        pass
    return None

def run():
    from playwright.sync_api import sync_playwright
    
    result = {"success": False, "api_key": None, "error": None, "step": "start"}
    
    with sync_playwright() as p:
        print("Launching headless Chromium...")
        browser = p.chromium.launch(
            headless=True,
            args=['--no-sandbox', '--disable-setuid-sandbox']
        )
        
        # Use persistent context with copy of existing profile
        context = browser.new_context(
            user_agent='Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36'
        )
        
        page = context.new_page()
        result["step"] = "browser_ready"
        
        print("Navigating to Tinker...")
        try:
            # Try direct access first (maybe session cookie still valid)
            page.goto("https://tinker-console.thinkingmachines.ai/keys", timeout=15000)
            page.wait_for_load_state("networkidle", timeout=10000)
            current_url = page.url
            print(f"Current URL: {current_url}")
            result["step"] = f"navigated: {current_url}"
            
            # Check if we're on auth page
            if "auth.thinkingmachines.ai" in current_url or "sign" in current_url.lower():
                print("Need to authenticate...")
                result["step"] = "need_auth"
                
                # Try Google OAuth
                page.goto("https://auth.thinkingmachines.ai/api/login?provider=GoogleOAuth&redirect_uri=https%3A%2F%2Ftinker-console.thinkingmachines.ai%2Fcallback&client_id=client_01JT41MFTDNYP0RYJ9MF318GDF&source=signin", timeout=15000)
                page.wait_for_load_state("networkidle", timeout=10000)
                
                print(f"OAuth URL: {page.url}")
                
                # Google account chooser - type email
                if "accounts.google.com" in page.url:
                    try:
                        # Check if account chooser shows up
                        if page.locator("text=muchammadizzuddin@gmail.com").count() > 0:
                            page.locator("text=muchammadizzuddin@gmail.com").click()
                        else:
                            # Email input
                            page.fill('input[type="email"]', "muchammadizzuddin@gmail.com")
                            page.click('button:has-text("Next")')
                            page.wait_for_load_state("networkidle", timeout=10000)
                            
                            # If password needed
                            if page.locator('input[type="password"]').count() > 0:
                                result["error"] = "Password required for Google login"
                                result["step"] = "need_google_password"
                                print("BLOCKED: Need Google password")
                                context.close()
                                browser.close()
                                with open(OUTPUT_FILE, 'w') as f:
                                    json.dump(result, f, indent=2)
                                return result
                        
                        page.wait_for_load_state("networkidle", timeout=10000)
                        print(f"After Google auth: {page.url}")
                    except Exception as e:
                        print(f"Google auth error: {e}")
                        result["error"] = str(e)
            
            # Check if on onboarding
            if "onboarding" in page.url:
                print("On onboarding page, completing...")
                result["step"] = "onboarding"
                
                # Accept ToS if not accepted
                tos_checkbox = page.locator('input[type="checkbox"]')
                if tos_checkbox.count() > 0 and not tos_checkbox.is_checked():
                    tos_checkbox.check()
                
                # Click Continue
                continue_btn = page.locator('button:has-text("Continue")')
                if continue_btn.count() > 0 and continue_btn.is_enabled():
                    continue_btn.click()
                    page.wait_for_load_state("networkidle", timeout=10000)
                    print(f"After onboarding: {page.url}")
            
            # Now try to get to keys page
            if "keys" in page.url or page.locator("text=API Key").count() > 0 or page.locator("text=key").count() > 0:
                print("On keys page!")
                result["step"] = "keys_page"
                
                # Grab page content
                content = page.content()
                
                # Look for API key patterns
                import re
                keys = re.findall(r'["\']?(tm_[a-zA-Z0-9_-]{20,})["\']?', content)
                if not keys:
                    keys = re.findall(r'["\']?([a-zA-Z0-9_-]{32,})["\']?', content)
                
                print(f"Found potential keys: {keys[:3]}")
                result["api_key"] = keys[0] if keys else None
                result["page_content_preview"] = content[:2000]
                result["success"] = True
            else:
                # Save page state
                result["final_url"] = page.url
                result["page_title"] = page.title()
                html = page.content()
                result["page_preview"] = html[:3000]
                
        except Exception as e:
            result["error"] = str(e)
            print(f"Error: {e}")
        
        context.close()
        browser.close()
    
    with open(OUTPUT_FILE, 'w') as f:
        json.dump(result, f, indent=2)
    
    print(f"\n=== RESULT ===")
    print(json.dumps(result, indent=2))
    return result

if __name__ == "__main__":
    run()
